library/std/src/sys/pal/unix/stack_overflow.rs

```diff
@@ -116,32 +123,43 @@ mod imp {
     static MAIN_ALTSTACK: AtomicPtr<libc::c_void> = AtomicPtr::new(ptr::null_mut());
     static NEED_ALTSTACK: AtomicBool = AtomicBool::new(false);
 
+    /// # Safety
+    /// Must be called only once
+    #[forbid(unsafe_op_in_unsafe_fn)]
     pub unsafe fn init() {
         PAGE_SIZE.store(os::page_size(), Ordering::Relaxed);
 
         // Always write to GUARD to ensure the TLS variable is allocated.
-        let guard = install_main_guard().unwrap_or(0..0);
+        let guard = unsafe { install_main_guard().unwrap_or(0..0) };
         GUARD.set((guard.start, guard.end));
 
-        let mut action: sigaction = mem::zeroed();
+        // SAFETY: assuming all platforms define struct sigaction as "zero-initializable"
+        let mut action: sigaction = unsafe { mem::zeroed() };
         for &signal in &[SIGSEGV, SIGBUS] {
-            sigaction(signal, ptr::null_mut(), &mut action);
+            // SAFETY: just fetches the current signal handler into action
+            unsafe { sigaction(signal, ptr::null_mut(), &mut action) };
             // Configure our signal handler if one is not already set.
             if action.sa_sigaction == SIG_DFL {
+                if !NEED_ALTSTACK.load(Ordering::Relaxed) {
```

Comment:
This can't be true in the first loop iteration, `NEED_ALTSTACK` is only set to true here and the safety condition ensures this is only called once. For the second loop, this is wrong, since the `SIGBUS` handler will never be registered if the `SIGSEGV` handler has been installed correctly.

Reply:
@joboet This "handler" type only covers the sigaltstack.

Reply:
- This condition _only_ covers the sigaltstack.
- This condition _only_ prevents installing the sigaltstack twice.

...actually, so does the code in `make_handler`, now that I look at it closely.